### ISSUE TITLE
Gaybar 1.5

### ISSUE THREAD

**Issue by josephbirkner** [Roboy/ravestate#20]
### About
Gaybar is an algorithm for matching aging picky signal instances (__Spikes__) with picky State activations, modeled as a bipartite bidirectional graph `G_Bar = {I, A, E}`:
* I is a set of signal instances
* A is a set of unfulfilled state activations
* E are edges, where `E ⊑ {(i, a) | i in I, a in A}`.

### Algorithm
Eager state activations are sitting at the bar, waiting for matching Spikes. They have criteria wrt/ the Spikes they wish to run their state activation function with, including...
* ... the spike's signal's name,
* ... the amount of time the spike has spent in the bar (it's _age_),
* the quantity of instances.

Spikes entering the bar wish to activate as many attractive states as possible, for as many different context properties as possible. A state is considered more attractive, the more specific it's criteria are, promising a higher reduction of local entropy in the bar. The limiting constraint is given by the context properties that are written to by a state: _A Spike may only activate one writing state for every context property_. Here is the formula for calculating the exact attractiveness of a state, wrt/ each of it's possible disjunct constraint conjunctions:

<a href="https://www.codecogs.com/eqnedit.php?latex=\text{Attr}_{State}=&space;\begin{Bmatrix}&space;\left&space;(&space;\sum_{s&space;\in&space;c}&space;\frac{1}{\text{num\_suitors}(s)}&space;\right&space;)_{c&space;\in&space;\text{Conjunct}_{State}}&space;\end{Bmatrix}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\text{Attr}_{State}=&space;\begin{Bmatrix}&space;\left&space;(&space;\sum_{s&space;\in&space;c}&space;\frac{1}{\text{num\_suitors}(s)}&space;\right&space;)_{c&space;\in&space;\text{Conjunct}_{State}}&space;\end{Bmatrix}" title="\text{Attr}_{State}= \begin{Bmatrix} \left ( \sum_{s \in c} \frac{1}{\text{num\_suitors}(s)} \right )_{c \in \text{Conjunct}_{State}} \end{Bmatrix}" /></a>

As the spikes enter the bar, they approach interested activations. The activations take note of Spikes which match their criteria, and notify the Spikes about their desire. As soon as a state activation X has gathered a sufficient number of spikes, it proceeds to tell the lucky candidate spikes, which in turn tell competing activations about their impending non-availability. If no spike more attractive than X is interested in the spike, it will consent to the activation. Otherwise, it won't: X will need to wait until it is the most attractive remaining suitor. As such, responsibility to determine whether the future fulfillment of an activation's constraints is likely lies entirely on the activation, and it is *necessary* for an activation to reject a spike when it sees it's hopes for activation deminishing. This process is called __auto-elimination__.

Once all spikes have consented, X runs it's respective state's activation function. Henceforward, X's partaking Spikes will no longer associate themselves with state activations that would write to context properties such as the those they just saw (let's just say: Spikes are always looking for new experiences).

In the bar, time is of the essence though; asynchronously and impartially progressing at a certain number of ticks per second. With each tick, Spikes gain in age. As ticks pass for a Spike, new suitable candidate activations may become available for it, while previous activations may lose interest. A Spike finally leaves the bar, when no activations are interested in it anymore.

### Spike Offspring and the Causal Group
As a state's activation function runs, it may trigger a number of offspring spikes (property:changed, state:fancy-signal, etc.). It is grave here, that whichever Spikes _caused_ this activation be notified about these offspring spikes, such that the offspring and parents synchronize wrt/ to their written properties: Offspring will not write to properties that were written by any parent, and parents will not write to properties that were written by any child. Such a set of causally linked spikes is called a __causal group__.

E.g. as a consequence, a `verbalizer:intent:changed` spike indirectly caused by some `rawio:in:changed` is _not_ going trigger a write to `rawio:out`, if the same `rawio:in:changed` already triggered a write to `rawio:out` through a different activation path:

**TLDR:** Causally connected spikes share a causal group, and as such have identical unwritten properties.
